tsconfig.json

```diff
@@ -0,0 +1,11 @@
+{
+  "compilerOptions": {
+    "noImplicitAny": false,
+    "allowJs": true,
```

Comment:
This is important for a step-by-step migration

Reply:
This attempts to migrate the whole project though, in the sense that it pipes everything through the TS compiler. 
I think it'll be easier to manage if we only run TS on modules/adapters that choose to use TS. It may be tricky to set up configuration wise but it should be possible, in the same way that a TS project can use JS libraries from npm. 

What I'm thinking is the compilation (and most of the codebase, including all of core) should be left as it is now; except for an additional step in the beginning that converts any `modules/**/*.ts` into JS.

Reply:
I'm guessing that piping this through `tsc` may have unintentional side effects. And I agree on that. However it is far less complex in configuration and mentally. If `tcs` outputs bascially the js files as is, because the `target` and `module` is set to the system prebidjs uses, they should bascially look the same.

Tinkering with [--noEmit](https://www.typescriptlang.org/tsconfig#noEmit) or babel-tyepscript preset, fiddling in typechecking at a later point, is IMHO not worth the effort.

The worst thing that can happen is that typescript generates additional code that would not be required, because babel is creating it. If the `import` statements are left intact, the hardest part still works (module resolution).

> What I'm thinking is the compilation (and most of the codebase, including all of core) should be left as it is now; except for an additional step in the beginning that converts any `modules/**/*.ts` into JS.

I agree with you, but I try to configure `tsc` that it does not alter the `js` files and just pipes them through.

Reply:
IMO doing this invites treating the whole project as TS, or maybe even requires it - I see have already started with migrating `prebid.ts`. The fact that it seems to mean adding useless "any" declarations is what worries me.

My approach would be to move all ts files into a separate tree and have tsc work only on that, then merge its output with "normal" sources and have the current build start there. I'd expect this to be easier than trying to police tsc in what it does with the non-ts portion of the project; all it requires is a way to tell it "here's where you find core source". We don't even necessarily need to resolve relative imports; since all ts files would be new, we could have them do `import {registerBidder} from 'core/adapters/bidderFactory'` instead of `from '../src/adapters/bidderFactory'`.

Reply:
I'll add that if you can do this without having to add the `any` tags in core, I'll be a lot less skeptical :)

Reply:
TL;DR: `any` is temporary. `noImplicitAny` compiler option is the goal 

`any` is a crutch that will be removed. I actually had to add this for `Object.entries` because the - [ ] type is `[string, unknown]` . Unknown is like any, but you have to check everything beforehand.

This is very sensible as you have no idea what properties you are enumerating. I'm forcing typescript to believe that the devs exactly know what they are doing.🙈

The path forward is to add proper types.

I started with `prebid.js` , because this is the only hardcoded file in the gulp build 😁

Reply:
I managed to reduce it to the `ts` files and the output looks nearly identical to the javascript before.

Due to the complexity of the build I'm wrapping my head around, this can definetly be a crutch until all js files are ported to ts. Then we may refactor the build a little built to perform linting only on ts files